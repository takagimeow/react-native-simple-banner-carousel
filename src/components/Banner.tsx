import React from 'react';
import { Image, TouchableOpacity, View } from 'react-native';

export interface IBanner {
  id: string;
  source: any;
  width: number;
  onPress?: (id: string) => void;
}

export function Banner({ id, width, source, onPress }: IBanner) {
  return (
    <View key={id}>
      <TouchableOpacity onPress={onPress ? () => onPress(id) : undefined}>
        <Image
          source={source}
          style={{
            width,
            maxWidth: '100%',
            height: 156,
            backgroundColor: '#bbb',
            borderRadius: 10,
          }}
        />
      </TouchableOpacity>
    </View>
  );
}
