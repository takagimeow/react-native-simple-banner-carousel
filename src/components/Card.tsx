import React from 'react';
import { Image, StyleProp, Text, View, ViewStyle } from 'react-native';

export interface ICard {
  id: string;
  title: string;
  width: number;
  source: any;
  description: string;
  score: number;
  onPress: () => void;
  style?: StyleProp<ViewStyle>;
}

export function Card({ id, source, width, title, description, score, style }: ICard) {
  return (
    <View
      key={id}
      style={[
        {
          width: width,
          maxWidth: '100%',
          backgroundColor: '#fffffff',
          paddingTop: 12,
          paddingBottom: 16,
        },
        style,
      ]}
    >
      <Image
        source={source}
        style={{
          width: width,
          maxWidth: '100%',
          height: 156,
          backgroundColor: '#bbb',
          marginBottom: 8,
        }}
      />
      <View
        style={{
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
        }}
      >
        <Text
          style={{
            flex: 11,
            fontSize: 14,
          }}
        >
          {title.length > 25 ? `${title.slice(0, 25)}...` : title}
        </Text>
        <View
          style={{
            flex: 1,
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 25,
            paddingVertical: 6,
            paddingHorizontal: 4,
            backgroundColor: '#f0f0f0',
          }}
        >
          <Text
            style={{
              fontSize: 12,
            }}
          >
            {score}
          </Text>
        </View>
      </View>
      <View
        style={{
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
        }}
      >
        <Text
          style={{
            fontSize: 12,
          }}
        >
          {description}
        </Text>
      </View>
    </View>
  );
}
