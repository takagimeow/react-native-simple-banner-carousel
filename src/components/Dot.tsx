import { Platform, StyleProp, View, ViewStyle } from 'react-native';
import React, { useEffect, useState } from 'react';

export interface IDot {
  index: number;
  currentIndex: number;
  lastIndex: number;
}

const initialStyle = {
  borderRadius: 25,
  width: 6,
  height: 6,
  backgroundColor: '#f0f0f0',
  marginHorizontal: 2,
};
export function getStyle({ index, currentIndex, lastIndex }: IDot): StyleProp<ViewStyle> {
  if (Platform.OS === 'android') {
    return {
      ...initialStyle,
      backgroundColor: index === currentIndex ? '#ffffff' : 'transparent',
      borderWidth: index === currentIndex ? 0 : 0.7,
      borderColor: index === currentIndex ? 'transparent' : '#ffffff',
      marginHorizontal: 2,
    };
  } else {
    return {
      ...initialStyle,
      backgroundColor:
        index === currentIndex
          ? index === 0 || index === lastIndex
            ? '#ffffff'
            : '#000000'
          : '#f0f0f0',
      opacity: index === currentIndex ? 1 : 0.5,
      width: 8,
      height: 8,
      marginHorizontal: 5,
    };
  }
}

export function Dot({ index, currentIndex, lastIndex }: IDot) {
  const [style, setStyle] = useState<StyleProp<ViewStyle>>(
    getStyle({ index, currentIndex, lastIndex }),
  );
  useEffect(() => {
    setStyle(getStyle({ index, currentIndex, lastIndex }));
  }, [index, currentIndex]);
  return <View style={style} />;
}
