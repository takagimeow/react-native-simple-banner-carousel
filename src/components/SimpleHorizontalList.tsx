import React, { useEffect, useRef, useState } from 'react';
import { Animated, Dimensions, Platform, View } from 'react-native';

const MARGIN = 8;
const INSET = -15;

export interface ISimpleHorizontalList<T> {
  data: T[];
  renderItem: (props: T, index: number, itemWidth: number) => React.ReactElement;
}

export function SimpleHorizontalList<T>({ data, renderItem }: ISimpleHorizontalList<T>) {
  const scrollX = useRef(new Animated.Value(0)).current;
  const [itemWidth] = useState(Dimensions.get('screen').width * 0.75);
  const [, setDataIndex] = useState(0);

  useEffect(() => {
    scrollX.addListener(event => {
      const newIndex = Math.round(Math.abs(event.value / itemWidth));
      setDataIndex(newIndex);
    });

    return () => {
      scrollX.removeAllListeners();
    };
  }, [scrollX]);

  return (
    <View
      style={{
        width: '100%',
      }}
    >
      <Animated.ScrollView
        horizontal
        style={{
          backgroundColor: 'transparent',
          padding: 0,
          margin: 0,
        }}
        contentContainerStyle={[
          {
            paddingHorizontal: Platform.OS === 'android' ? MARGIN + INSET : 0,
          },
        ]}
        contentInset={{
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
        }}
        contentOffset={{
          x: 0,
          y: 0,
        }}
        snapToInterval={itemWidth + MARGIN * 2}
        showsHorizontalScrollIndicator={false}
        scrollEventThrottle={16}
        snapToAlignment="start"
        decelerationRate="fast"
        pagingEnabled
        onScroll={Animated.event(
          [
            {
              nativeEvent: {
                contentOffset: { x: scrollX },
              },
            },
          ],
          {
            useNativeDriver: true,
          },
        )}
        bounces={false}
        alwaysBounceHorizontal={false}
        alwaysBounceVertical={false}
      >
        {data.map((item, i) => (
          <View
            key={`simple_carousel_item_key_${i}`}
            style={{
              margin: MARGIN,
            }}
          >
            {renderItem(item, i, itemWidth)}
          </View>
        ))}
      </Animated.ScrollView>
    </View>
  );
}
