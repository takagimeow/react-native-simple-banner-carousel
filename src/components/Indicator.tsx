import React from 'react';
import { Dimensions, View } from 'react-native';
import { Dot } from './Dot';

export interface IIndicator<T> {
  data: T[];
  currentIndex: number;
  lastIndex: number;
}

export function Indicator<T>({ data, currentIndex, lastIndex }: IIndicator<T>) {
  return (
    <View
      style={{
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
        position: 'absolute',
        bottom: '5%',
        left: 0,
        width: Dimensions.get('window').width,
        height: 24,
        backgroundColor: 'transparent',
        zIndex: 1,
      }}
    >
      {data.map((_, i) => (
        <Dot
          key={`simple_carousel_indicator_key_${i}`}
          index={i}
          currentIndex={currentIndex}
          lastIndex={lastIndex}
        />
      ))}
    </View>
  );
}
