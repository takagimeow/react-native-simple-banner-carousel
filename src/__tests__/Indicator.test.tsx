/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';
import 'react-native';
import { render } from '@testing-library/react-native';

import { Indicator } from '../components/Indicator';

describe('Indicator', () => {
  it('Snapshot Test', () => {
    const { toJSON } = render(
      <Indicator
        data={[
          {
            name: 'tanaka',
          },
          {
            name: 'suzuki',
          },
        ]}
        currentIndex={0}
        lastIndex={2}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
