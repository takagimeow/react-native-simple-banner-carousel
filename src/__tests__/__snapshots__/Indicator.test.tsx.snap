// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator Snapshot Test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "bottom": "5%",
      "display": "flex",
      "flexDirection": "row",
      "height": 24,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "width": 750,
      "zIndex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 25,
        "height": 8,
        "marginHorizontal": 5,
        "opacity": 1,
        "width": 8,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#f0f0f0",
        "borderRadius": 25,
        "height": 8,
        "marginHorizontal": 5,
        "opacity": 0.5,
        "width": 8,
      }
    }
  />
</View>
`;
